numpy/distutils/clang_cl.py

```diff
@@ -0,0 +1,167 @@
+                self.compile_options.remove(opt)
+        ltcg = "/LTCG"
+        ldflags = self._ldflags
+        for key in ldflags:
+            if ltcg in ldflags[key]:
+                ldflags[key].remove(ltcg)
+        self.compile_options.extend(
+            [
+                "/O2",
+                "/GS-",
+                "-Wno-visibility",
+                "-Wno-logical-op-parentheses",
+                "-Wno-microsoft-include",
+                "-Wno-shift-op-parentheses",
+            ]
+        )
+
+        # Merge current and previous values of 'lib' and 'include'
+        os.environ["lib"] = _merge(environ_lib, os.environ["lib"])
+        os.environ["include"] = _merge(environ_include, os.environ["include"])
+        # Get version information about clang which is needed to find
+        # the correct include and lib directories
+        clang_base = os.path.split(self.cc)[0]
+        clang_version_cmd = [self.cc, "--version"]
+        # Get version or let the user know if there was an issue
+        try:
+            clang_full_ver = subprocess.check_output(
+                clang_version_cmd
+            ).decode()
+        except WindowsError:
+            raise WindowsError(
+                "clang-cl.exe could not be found. It must be on your path."
+            )
+        clang_version = ""
+        if "version" in clang_full_ver:
+            loc = clang_full_ver.find("version")
+            clang_ver_string = clang_full_ver[loc + 7:]
+            clang_version = clang_ver_string.split("\n")[0].strip()
+        clang_command_and_return = (
+            f"Running the command \n\n{' '.join(clang_version_cmd)}\n\n "
+            f"which returned \n\n{clang_full_ver}"
+        )
+
+        if not clang_version:
+            raise RuntimeError(
+                f"The clang version could not be detected from the version "
+                f"string. {clang_command_and_return}"
+            )
+
+        # Forbid compiling 32 on 64-bit LLVM and vice versa
+        target = "i686" if platform_bits == 32 else "x86_64"
+        if target not in clang_full_ver:
+            raise RuntimeError(
+                f"clang-cl must target {target} when building on "
+                f"{platform_bits}-bit windows. {clang_command_and_return}"
+            )
+        # Add the include directory for clang
+        clang_incl = ["..", "lib", "clang", clang_version, "include"]
+        clang_incl = os.path.abspath(os.path.join(clang_base, *clang_incl))
+        if not os.path.exists(clang_incl):
+            raise RuntimeError(
+                f"{clang_incl} could not be found. This directory contains "
+                f"includes required by clang-cl."
+            )
+        self.include_dirs.insert(0, clang_incl)
+
+        # Need to block MS includes except those from Visual Studio 14.0
+        # since they include intrinsics that break clang
+        def retain_include(path):
+            blocked = ("Windows Kits", "MSVC", "2019")
```

Comment:
Is it guaranteed that the path for Visual Studio 14.0 doesn't include "MSVC"?

Reply:
This is definitely the most challenging part of this.  I've linked to a SO question that had the relevant details:

https://stackoverflow.com/questions/36086328

Reply:
Thanks, very useful discussion.